src/operator/controllers/intents_reconcilers/database/database_reconciler.go

```diff
@@ -76,87 +89,150 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, errors.Wrap(err)
 	}
 	username := clusterutils.BuildHashedUsername(clientIntents.GetServiceName(), clientIntents.Namespace, clusterID)
-	pgUsername := clusterutils.KubernetesToPostgresName(username)
-	intentsDeleted := !clientIntents.DeletionTimestamp.IsZero()
-	action := lo.Ternary(intentsDeleted, otterizev1alpha3.DBPermissionChangeDelete, otterizev1alpha3.DBPermissionChangeApply)
+	dbUsername := clusterutils.KubernetesToPostgresName(username)
 
-	// If intents are being deleted, skip directly to permission cleaning
-	if !intentsDeleted {
-		for databaseInstance, intents := range dbInstanceToIntents {
-			pgServerConf, err := findMatchingPGServerConfForDBInstance(databaseInstance, pgServerConfigs)
-			if err != nil {
-				r.RecordWarningEventf(clientIntents, ReasonMissingPostgresServerConfig,
-					"Could not find matching PostgreSQLServerConfig. Error: %s", err.Error())
-				return ctrl.Result{}, nil // Not returning error on purpose, missing PGServerConf - record event and move on
-			}
-			pgConfigurator := postgres.NewPostgresConfigurator(pgServerConf.Spec)
-			dbnameToDatabaseResources := getDBNameToDatabaseResourcesFromIntents(intents)
-			err = pgConfigurator.ConfigureDatabasePermissions(ctx, pgUsername, action, dbnameToDatabaseResources)
-			if err != nil {
-				r.RecordWarningEventf(clientIntents, ReasonApplyingDatabaseIntentsFailed,
-					"Failed applying database clientIntents: %s", err.Error())
-				return ctrl.Result{}, errors.Wrap(err)
-			}
+	dbIntents := clientIntents.GetDatabaseIntents()
+	dbInstanceToIntents := lo.GroupBy(dbIntents, func(intent otterizev1alpha3.Intent) string {
+		return intent.Name // "Name" is the db instance name in our case.
+	})
+
+	existingPGInstances := lo.Map(pgServerConfigs.Items, func(config otterizev1alpha3.PostgreSQLServerConfig, _ int) string {
+		return config.Name
+	})
+	existingMySQLInstances := lo.Map(mySQLServerConfigs.Items, func(config otterizev1alpha3.MySQLServerConfig, _ int) string {
+		return config.Name
+	})
+
+	existingDBInstances := append(existingPGInstances, existingMySQLInstances...)
+
+	missingDBServerConfig := lo.Without(lo.Keys(dbInstanceToIntents), existingDBInstances...)
+	for _, missingDBInstance := range missingDBServerConfig {
+		// Not returning error on purpose, missing PGServerConf - record event and move on
+		// When a new PGServerConf is created, the clientIntents will be reconciled again
+		r.RecordWarningEventf(clientIntents, ReasonMissingDBServerConfig,
+			"Missing database server config: did not find DB server config to match database '%s' in the cluster", missingDBInstance)
+	}
+
+	for _, config := range pgServerConfigs.Items {
```

Comment:
Wouldn't missing server configs still be iterated here, causing an error ?

Reply:
no. 
`missingDBServerConfig` contains DB names referred to by intents, where the db name doesn't have a matching postgres/mysqlServerConfig. We won't iterate over them because they're not in `pgServerConfigs` / `mySQLServerConfigs`. 
We will (correctly) iterate over dbServerConfigs with no intents, to properly trigger cleanup of instances with no applied intents.

Reply:
I'll rename some vars to make it clearer.